### ISSUE TITLE
Allow to import system fonts from within Godot

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#306]
**Describe the project you are working on:**
This applies to numerous projects, all using GUI.

**Describe the problem or limitation you are having in your project:**
Every time I create a new project that uses GUI, I inevitably fall into the same problem - I need a font. Now, I'm used to needing to find a TTF file and import it into project so that I can put some labels. But I've come into _some engines_ that handle this in better way which is, most importantly, more friendly towards beginners. Making labels of different size has multiple problems. Scaled labels/buttons look super-ugly, but aside from that, their anchors are broken (https://github.com/godotengine/godot/issues/19068). So using a dynamic font for text size is a necessity.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
The biggest problem with this workflow is whenever you want to have nice text, you have to go out of your way and get a TTF file from somewhere. It might be web or system fonts, but I imagine this is not obvious step for beginners.

I've seen that in _some other engines_, the fonts are seemingly built-in. Not sure if they are provided with the engine (which is a no-no-no for Godot), but the option to get a system font into your project without having to navigate to Windows/Linux directory would be really useful.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![image](https://user-images.githubusercontent.com/2223172/70855414-b2c98600-1eca-11ea-8b96-ad21bd254c12.png)
(imagine there's some button saying "Import system font" here and clicking it opens some dropdown/lists of system fonts, clicking which will copy the font into your project)

**Describe implementation detail for your proposal (in code), if possible:**
I don't recall any similar system in Godot, so can't recommend any copy-paste material here, but the idea is to access system's font directory (like C:/Windows/Fonts or /Linux/Fonts) and allow to copy a font into your project without leaving the editor. So it requires platform-specific code.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
This will be used by everyone the second this is added, trust me.

**Is there a reason why this should be core and not an add-on in the asset library?:**
The purpose is to be friendly to newbies, so having this in asset lib and being basically a "standard plugin to use" would make it a bit pointless...

**Comment by golddotasksquestions** [comment]
Being able to also load systemfonts directly would be great!

As a beginner, the biggest hurdle to overcome for me was to find the place where I should load the actual font file. Still to this day I have to "look for" the place a few seconds if I have not used Richlabel or Labels in a while. The other hurdle for beginners is that if they found the place "Custom Font", they need two go down two more sub-stages until they finally can load the font. They need to select "Dynamic Font", then find "Font" and only under "Font Data" they are finally at the place.

As UI designers we have to keep in mind every click into a sub menu is a confidence check. If the user thinks this is a very popular menu item, they expect it to be very prominent and very surface level. If they won't find it on the surface level, they will as themselves with every click: am I still at the right place? Is this where I should look?

**Comment by Calinou** [comment]
I'm not sure about this, as Windows' and macOS' system fonts are under proprietary licenses. You definitely don't want to include them in your exported PCK :slightly_smiling_face:

That said, *loading* system fonts without including them in the PCK could be nice for GUI applications.

**Comment by volzhs** [comment]
yes. allowing load system font will be great so that game can show any character not embedded.

**Comment by paulhocker** [comment]
It is a best practice to have an asset be part of the project and not part of the operating system it is running on. Relying on a system font (or any file) could lead to confusion when moving or sharing a project on Windows and Mac for example. I will be down voting this.

**Comment by KoBeWi** [comment]
> Relying on a system font (or any file) could lead to confusion when moving or sharing a project on Windows and Mac for example. I will be down voting this.

lol, but I mean copying the font for the project, so it becomes an asset. It's great for prototyping and stuff.

**Comment by paulhocker** [comment]
I see -- sorry I missed that in the original post. Changing my mind to up vote.

**Comment by Calinou** [comment]
@paulhocker This is actually even worse when you're distributing the project publicly, as you'll most likely be violating the font author's EULA :slightly_smiling_face:

**Comment by bruvzg** [comment]
> @paulhocker This is actually even worse when you're distributing the project publicly, as you'll most likely be violating the font author's EULA when doing so.

And it will be confusing for many. Users will select their legitimately bought fonts, like they usually do in their graphic design tools, without even thinking that text won't be rendered to image on export, and they actually need separate embedding licences.

**Comment by paulhocker** [comment]
> 
> 
> > @paulhocker This is actually even worse when you're distributing the project publicly, as you'll most likely be violating the font author's EULA when doing so.
> 
> And it will be confusing for many. Users will select their legitimately bought fonts, like they usually do in their graphic design tools, without even thinking that text won't be rendered to image on export, and they actually need separate embedding licences.

That is already a problem today, so I don't see how this request makes it worse. It is the responsibility of the developer to make sure they are following the license for assets in their final build. That said, I think this idea would be a nice to have, as there are so many other things to work on.

**Comment by golddotasksquestions** [comment]
If you bought your OS, you also bought the licensing to use the fonts that come with it for whatever purpose other than redistribution. As long as you are just redistributing a project that is using the font, not the font itself, there is absolutely no problem using the system fonts of an OS you own in your project.

**Comment by Zireael07** [comment]
The problem is, if the font is simply copied into the project, it gets very easy for people not to notice that they're redistributing (and if the font is in the project, it is redistributing, plain and simple)

**Comment by golddotasksquestions** [comment]
Well of course it needs to be embedded, not just added to the zip the user has to unpack to play the game.
I have never tried to dismantle a pck file. But since you also can extract textures models sounds and other resources from other games without too much difficulty, I would assume there would not be much difference to being able to extract a font from a pck, legally. I'm not a lawyer though.

**Comment by Zireael07** [comment]
To be honest, I cannot recall an engine that would do what the original requester said ("fonts are seemingly built-in") without also redistributing them. Unreal and Unity's default font are obviously bundled with the engine :)

And redistributing is not about whether the pck can be dismantled or not, if the font file is there, you are redistributing it. The only reasonable alternative is for the engine to somehow detect the use of system fonts and automatically attempt to use system font on the end-user's computer. Again, I can't recall an engine that would do this.

**Comment by golddotasksquestions** [comment]
It's always possible to reconstruct or reverse engineer a font, even from a analog book. The questions is what the law considers enough reasonable effort to prevent that from happening.

**Comment by Calinou** [comment]
@golddotasksquestions Font files are widely considered to be copyrightable by now. Either way, this is getting off-topic.

**Comment by KoBeWi** [comment]
Hmmm, another option could be putting some free fonts in AssetLib and making them easily accessible. That way we get good font support and avoid copyright issues.

**Comment by Calinou** [comment]
@KoBeWi That can be done, but I don't see much point since downloadable font repositories are readily available (such as [google/fonts](https://github.com/google/fonts)).

**Comment by KoBeWi** [comment]
Well, it could be any source. It just should be easy to get from within the editor, while avoiding embedding it with the engine (which reminds me that the engine itself has a nice font that could be optionally saved as resource within the project...).

**Comment by golddotasksquestions** [comment]
@Calinou of course fonts are copyrighted. If they were not, you would not need to obtain a license to use them. Why do you say my comments are off topic?

**Comment by dalexeev** [comment]
https://github.com/godotengine/godot/issues/32916#issuecomment-544170924

**Comment by Calinou** [comment]
In case anyone is interested on working on this, note that Windows 10 1809 onwards installs custom fonts into an user-specific folder. See https://gitlab.com/inkscape/inkscape/issues/50 for more information.

**Comment by follower** [comment]
> another option could be putting some free fonts in the AssetLib

I discovered sometime last year that someone has actually already done that:

 * "[Open Font Package](https://godotengine.org/asset-library/asset/316)" on the Godot Asset Library

It includes a selection of fonts including Droid Sans, Source Code Pro & some more novelty ones.

Repository here: https://github.com/dalton5000/Godot-Fontpack/

It notes the following which I think is why having a Godot-specific solution is helpful--particularly for beginners who may not even be aware of other font sources:

> "Each font folder contains the .tres file for Godot, additional font versions if they available and its license file."

**Comment by Calinou** [comment]
@follower The font size is defined in the `.tres` file, so newcomers would still have to edit the font resource to suit their needs.

**Comment by Shazoo** [comment]
allowing use the system built-in fonts will be great and necessary for CJK application  . 

A lot of Chinese fonts are almost 20M (file size) . it enlarge the game distribution size .

**Comment by Calinou** [comment]
Now that https://github.com/godotengine/godot/pull/39603 has been merged, there isn't much of a point to load system fonts for prototyping purposes.

However, the point above (loading system CJK/emoji fonts to reduce exported project file size) still stands. I was thinking we should have a **Load System Font** enum property in DynamicFont with several possible values. Since this implies loading different fonts depending on the target platform, I feel consolidating a few popular choices within an enum makes sense here. This way, you don't have to think about which fonts are guaranteed to be present on target platforms.

- None
- System Sans
  - Use Segoe UI on Windows, San Francisco on macOS and iOS, Roboto/Ubuntu/(Arial fallback) on Linux, Roboto on Android.
- System Serif
  - Use the system serif font. This is the least important option, so we don't *have* to implement this one.
- System CJK
  - Use the system-provided CJK font.
- System Emoji
  - Load the system emoji font.

When the **Load System Font** property is set to a value other than None, the font path is ignored (and hidden in the inspector).

The documentation should mention that font metrics *will* vary slightly across platforms, leading to possible UI layout inconsistencies.

**Comment by Sosasees** [comment]
In Android and Linux you can change the system font.

Since many people choose a non-default font, importing that font
instead of always having to hope that the game font is readable
is great for readability.

This is My personal reason to want this feature:
I am usually concerned for accessibility in my games.

**Comment by Calinou** [comment]
> In Android and Linux you can change the system font.
> 
> Since many people choose a non-default font, importing that font instead of always having to hope that the game font is readable is great for readability.
> 
> This is My personal reason to want this feature: I am usually concerned for accessibility in my games.

While this point is valid for non-game applications, it's very uncommon for *games* to use the system font by default. If you want to supply accessible fonts as an option to replace your game's standard font (such as Open Dyslexic), this is already feasible :slightly_smiling_face:

**Comment by Sosasees** [comment]
> While this point is valid for non-game applications, it's very uncommon for games to use the system font by default.

i actually have a game planned where i don't have a thematic reason to use a specific font, so using the system font would be good for this game.

and even if i end up using a custom font:
i believe that Godot will become a frequent choice for Application development (not just game development) when Godot 4 releases, so not having to redownload the Godot font with every Godot-made application will save very much save disk space on many computers.

> If you want to supply accessible fonts as an option to replace your game's standard font (such as Open Dyslexic), this is already feasible 🙂

Thanks for suggesting accessible fonts.

**Comment by mhilbrunner** [comment]
I agree that this makes it too easy to think just copying and distributing .ttf files from your fonts directory is not a murky legal problem you need to think about.

How about we provide a nice set of default fonts for DynamicFont in Godot by default, and have checkboxes for the Project to set which charsets/languages the project needs to support, and then set up the default font(s) accordingly?
I _guess_ it would also be cool to have per language/locale exports or somesuch, but not sure.

Anyway, adding "one click import from C:\Windows\Fonts\" is somewhat akin to "one click import from Google Image Search to Texture2D", in that sure, in the end its the users responsibility and you can always say "of course you need proper licenses, why didn't you think of that, only use it with searches for creative commons/freely licenses stuff, OBVIOUSLY!", but... meh.

**Comment by Zireael07** [comment]
> How about we provide a nice set of default fonts for DynamicFont in Godot, and have checkboxes for the Project to set which charsets or languages the project needs to support, and then set up the default font(s) accordingly?

It would also be cool to have per language/locale exports or somesuch, but not sure.

I think there's a related issue about decreasing the binary sizes by removing some fonts (but search-fu is failing me), so it's either this or that.
Personally, I think they should be optional add-ons, we can support the majority of European languages easily, and then add an Arabic and CJK font options on top for those users who need those.

**Comment by bruvzg** [comment]
>I think there's a related issue about decreasing the binary sizes by removing some fonts

https://github.com/godotengine/godot-proposals/issues/3421 I guess

> we can support the majority of European languages easily, and then add an Arabic and CJK font options on top for those users who need those.

The only big one is CJK (and may become even larger with https://github.com/godotengine/godot/pull/56558), current Arabic font is like 150KB, most of the other language fonts are also 50-200KB each in size, but when bundled with the full editor and class db translation, splitting it to the add-ons make sense.

> How about we provide a nice set of default fonts for DynamicFont in Godot, and have checkboxes for the Project to set which charsets or languages the project needs to support, and then set up the default font(s) accordingly?

Extracting embedded editor font (on font from the editor add-on if we go this way) to the file, and adding it to the PCK/project folder should be easy to do.

> It would also be cool to have per language/locale exports or somesuch, but not sure.

I guess an option to export localized remaps to the separate PCK might be useful, but this will require proper PCK patch loading implemented, which AFAIK is not done.
